### ISSUE TITLE
Google Search - "Suspended: access denied"

### ISSUE THREAD

**Issue by ZhenyaPav** [searxng/searxng#2750]
**Version of SearXNG, commit number if you are using on master branch and stipulate if you forked SearXNG**
`2023.9.11+f182abd6`

**How did you install SearXNG?**
Installed using official guide.

**What happened?**
Google search always returns [Suspended: access denied](https://searx.zhenyapav.com/stats?engine=google). Despite this, engine stats report 100% reliability. The issue goes away after restarting the searxng service, for a small amount of time, but then returns.

**How To Reproduce**
Go to [searx.zhenyapav.com](https://searx.zhenyapav.com/), search for anything. Notice that Google results are absent.

**Expected behavior**
The results should include the ones from google Google.

**Comment by return42** [comment]
The blocks at Google are caused because bot requests are passed from the SearXNG instance to google and google then recognizes them as bot requests .. and is then blocking your instance.

@ZhenyaPav did you activated the [link_token method](https://github.com/searxng/searxng/issues/2498#issuecomment-1590625541) of the bot protection? / I recommend.

> The issue goes away after restarting the searxng service,

How long a engine is suspended depends on the [`suspended_times`](https://docs.searxng.org/admin/settings/settings_search.html) ..

https://github.com/searxng/searxng/blob/f182abd6f8f1eac20d19c3e4b4c9800115f2a705/searx/settings.yml#L48-L60

A value of `SearxEngineAccessDenied: 86400` sec (24h) is IMO to high .. on my instance I reduced the value to 60 sec. without any drawback.

The  [link_token method](https://github.com/searxng/searxng/issues/2498#issuecomment-1590625541) needs some training to catch all the bot IPs .. on my instance this method is active since about 2 month or so .. in the meantime the method identified round about 

On my instance I have this bot blocking method active since 2 or 3 month and my instance has build up a list of round about 3.500 IPv$ a dresses from where bot request are come / my instance is only IPv4 what makes bot detection simpler (the range of IPs bots can use is smaller).

Since now only a few bots manage to pass my instance I also have few problems with google --> https://darmarit.org/searx/stats?engine=google

TL;DR;

- link_token method needs to be active for a (very) long time to detect and block bots
- the IP of the instance must cool down at the search engine (Google)
- `SearxEngineAccessDenied` should be reduced to be able to work again first

**Comment by ZhenyaPav** [comment]
Reducing `SearxEngineAccessDenied` seems to solve the issue.

**Comment by unixfox** [comment]
Reopening this issue in order to investigate more.

The major thing to check is to see if the IP address is completely blocked from Google OR it's just one endpoint that do not work anymore. We have to check on an IP address that is blocked if doing searches on Google with a browser still work or not.

**Comment by ZhenyaPav** [comment]
I don't think my instance is blocked, at least when I used the server as SOCKS proxy, google worked fine.

**Comment by return42** [comment]
> Reopening this issue in order to investigate more.

@unifox I think [bot blocking and a reduced suspend time](https://github.com/searxng/searxng/issues/2750#issuecomment-1715268897) is the solution for this issue / compare https://github.com/paulgoio/searxng/issues/54#issuecomment-1717894345

**Comment by unixfox** [comment]
> > Reopening this issue in order to investigate more.
> 
> @unifox I think [bot blocking and a reduced suspend time](https://github.com/searxng/searxng/issues/2750#issuecomment-1715268897) is the solution for this issue / compare [paulgoio/searxng#54 (comment)](https://github.com/paulgoio/searxng/issues/54#issuecomment-1717894345)

For low traffic instances it is maybe a good solution, but for popular instances like mine it is not. My instance was straight blocked (I have not checked through a browser) and the only solution was to change the IP address, waiting 4 hours did not change anything.

Anyway I have not yet found the time to look into that, I'll hopefully this weekend.

**Comment by unixfox** [comment]
First investigation:

![image](https://github.com/searxng/searxng/assets/4016501/4d899d91-3baa-431e-affe-a5c6d8444d29)

That's the first time I have seen this message.

Usually it says:

![image](https://github.com/searxng/searxng/assets/4016501/1c01aab8-c1a4-4412-b8e6-0273fb369cb0)

This is going to be difficult to solve unless the instance changes their IP address.

**Comment by unixfox** [comment]
I have done my research, and we can't solve this issue by easily change some code into the Google engine.

This is also affecting many real users around the world that are not using SearXNG:
- https://community.t-mobile.com/troubleshooting-38/google-blocking-search-52786
- https://www.reddit.com/r/techsupport/comments/163zur6/google_blocking_searches_for_unaddressed_abuse/?rdt=57290
- https://www.reddit.com/r/tmobileisp/comments/1614dn2/abuse_complaints_about_malicious_behavior_is/

I think we will need to instruct our searxng instances maintainers to change their IP address. Thankfully, this blockage is rare and seems to happen maybe once per 2 months.

I propose to print an error message into the logs so that instances maintainers are aware of the issue and tell them to change the IP address.

Additional solutions that could help:

- Have a CAPTCHA solver into SearXNG.
- Make SearXNG look like more of a browser. Previously discussed in https://github.com/searxng/searxng/issues/2515, they are ways to make SearXNG behavior similar to a browser.
- Improve the limiter situation of the public instances, discussed in https://github.com/searxng/searxng/discussions/2802
- Have the ability to choose which IP protocol is used to send the request. Usually nowadays, servers have both IPv4 and IPv6 connectivity. We could have a parameter that allow to switch between IPv4 or IPv6, effectively allowing the maintainer to use IPv4 if IPv6 is blocked.
- Disable Google by default as an engine. I have seen public instances unaffected by this blockage, and some of them it's probably because they do not have Google enabled by default.

**Comment by unixfox** [comment]
Looking at the public instances that are currently blocked by Google, I see that all of them do not have the new advanced bot limiter enabled. You can see it in /config.

I think I'm really hoping that the new enforcement of the bot protection reduce the weird requests by bots and thus Google start to work again on 80% of the public instances, then stay like this for the future.

From my experience, I also think that Google recognize SearXNG as just one bot sending too many requests from many different servers to their servers (TLS and HTTP2 fingerprint). So if we can lower the overall amount of unnecessary requests done to their server as a whole on all the public instances, this will mean that Google will be less likely to block the SearXNG public instances.

**Comment by unixfox** [comment]
I'm going to close this issue as since we doubled the effort on forcing public instances to enable our advanced bot protection, I have not found any public instance blocked by google and not using our built-in bot protection.

**Comment by unixfox** [comment]
Unfortunately the first case of a block by google even with the bot protection:

https://search.bus-hit.me/search?q=%21go%20ok&language=en-US&time_range=&safesearch=0&categories=none

![image](https://github.com/searxng/searxng/assets/4016501/6bc74c21-f068-4f28-88e4-4708fe2e4b9f)

**Comment by return42** [comment]
Not sure google block is caused by SearXNG / the engine is deactivated by default .. but there are also other services that might send requests to google --> https://translate.bus-hit.me/ 

We should wait 24h (86400sec) and have a look again

https://github.com/searxng/searxng/blob/b013cbb4cc47f224f3785019ab4e56eeb264944f/searx/settings.yml#L48-L51

**Comment by unixfox** [comment]
> Not sure google block is caused by SearXNG / the engine is deactivated by default .. but there are also other services that might send requests to google --> [translate.bus-hit.me](https://translate.bus-hit.me/)

bots can use bangs too... I feel like bots have found how to circumvent our bot protection, that was expected.

I did send a message on matrix to the owner for asking what is going on.

> translate.bus-hit.me

it's a different domain and service, from my experience at google the blocks and rate limits are per domain.  
`translate.google.com` VS `google.com`  
see the service translate still works.

> We should wait 24h (86400sec) and have a look again

This block is permanent, look at searx.space there are many instances in the same case, most do not have the bot protection. They have never been unblocked for the moment.

**Comment by return42** [comment]
>  look at searx.space there are many instances in the same case

Do I'm looking at the right place? I see only 3 or 4 instances having a red cross in the "google" column and none of them have a "access denied".  I have only found one instance that had a "access denied" (in the past) .. https://searx.mha.fi/stats?engine=google .. but this instance do need an update I think.

**Comment by unixfox** [comment]
![image](https://github.com/searxng/searxng/assets/4016501/acd322b6-972c-4931-bf47-6da240fc5baf)

**Comment by return42** [comment]
IDK why this column has blank fields .. but when I go to the instance and look into the engine stats or do some `!go foo` searches on the instance they have a reliability of 100% .. thats why I look on searx.space here in the google column:

![image](https://github.com/searxng/searxng/assets/554536/c785fd6e-c252-4dbc-9030-ec2bc85ba65a)

**Comment by return42** [comment]
> ![image](https://user-images.githubusercontent.com/4016501/276283881-acd322b6-972c-4931-bf47-6da240fc5baf.png)

@unixfox some (50%?) of them have a `searx.exceptions.SearxEngineAccessDeniedException` which will be fixed when .. gets merged:

- https://github.com/searxng/searxng/pull/2973
